### DISCUSSION TITLE
cypress within() command seems to misbehave

### DISCUSSION THREAD

**Discussion by zadigus** [cypress-io/cypress#16011] [Questions and Help] [answered]
Hi,

I am currently migrating from cypress 4.12.1 to 5.6.0. Until now, I had e.g. this code working fine:

```js
Cypress.Commands.add('getNbCartItemsFromIndicator', () => {
  cy.get('div[id="cart-indicator"]').first().within(() => {
    return cy.get('div[id="number-of-products"]').invoke('text').then(parseInt)
  })
})
```

I was able in version 4.12.1 to do this with the above command:

```js
cy.getNbCartItemsFromIndicator().then(cartItems => console.log(cartItems))
```

and that would've outputted the number written in the `div` with `id=number-of-products`.

Now, with version 5.6.0, the above code returns the `div` with `id=cart-indicator` back, i.e. not the text contained in its child `div` element with `id=number-of-products`, as before. If I get rid of the line of code 

```js
cy.get('div[id="cart-indicator"]').first().within(() => {
```

(and the line with the corresponding closing curly brace of course), then it works. I would like to keep the code with the call to `within` because it makes my test more robust. What should I do now to make it work correctly?

**Comment by bahmutov** [comment, marked as answer]
In version 5.4.0 we have fixed the `.within` command to always yield the original parent element

https://docs.cypress.io/guides/references/changelog#5-4-0

Thus trying to `return` or `cy.wrap` another value have no effect, see the examples in https://glebbahmutov.com/cypress-examples/7.1.0/commands/querying.html#yields-the-original-element

I have clarified this in the `.within` command documentation https://github.com/cypress-io/cypress-documentation/pull/3848

**Reply by cbruge** [reply]
Thank you for pointing me in the right direction!

**Reply by zadigus** [reply]
So @bahmutov how do I refactor my code to achieve my original goal? I want to make sure that I get some element inside of some other element, get the text inside of the former element and return it from my command.

I guess what I am wanting to do does not follow best practices. Instead, I should rather use data-cy / data-test css properties, shouldn't I?

**Reply by bahmutov** [reply]
Take a look at https://glebbahmutov.com/cypress-examples/7.1.0/recipes/return-value-from-within.html - you would use a variable in the same lexical scope, then wrap it after the `.within` callback is over

```js
Cypress.Commands.add('getMagicNumber', (selector, options = {}) => {
  Cypress._.defaults(options, { log: true })
  cy.log(`getMagicNumber from **${selector}**`)

  let magicNumber
  cy.get(selector, options)
    .within(options, () => {
      cy.get('.magic', options)
        .invoke(options, 'text')
        .then(parseInt)
        .then((x) => {
          magicNumber = x
        })
    })
    .then(() => {
      cy.wrap(magicNumber, options)
    })
})
cy.getMagicNumber('#parent', { log: false }).should('equal', 42)
```

**Reply by zadigus** [reply]
thank you very much for your help, I didn't know about your nice examples

**Reply by bahmutov** [reply]
:) I wrote the recipe just to answer your question, and now it becomes a good test that runs against Cypress to make sure it is still the right answer
